cmd/cmd_test.go

```diff
@@ -0,0 +1,315 @@
+			split("config add-user --username foobar")},
+		{"config-get-users",
+			split("config get-users user")},
+		{"config-get-users",
+			split("config get-users")},
+		{"config-get-users",
+			split("config get-users --detailed")},
+		{"neg-config-add-user-no-username",
+			split("config add-user")},
+		{"neg-config-add-user-no-password",
+			split("config add-user --username foobar")},
+
+		// Cleanup
+		{"cleanup",
+			split("config delete-user user")},
+		{"cleanup",
+			split("config delete-user user2")},
+	}
+
+	run(t, tests)
+}
+
+func TestLocalContext(t *testing.T) {
+	setup(t.Name())
+
+	tests := []test{
+		{"neg-config-delete-endpoint-no-name",
+			split("config delete-endpoint")},
+		{"config-add-endpoint",
+			split("config add-endpoint --address localhost --port 1433")},
+		{"config-add-user",
+			split("config add-user --username foobar")},
+		{"config-add-context",
+			split("config add-context --user user --endpoint endpoint --name my-context")},
+		{"config-delete-context-cascade",
+			split("config delete-context my-context --cascade")},
+		{"config-view",
+			split("config view")},
+		{"config-view",
+			split("config view --raw")},
+
+		{"neg-config-add-user-bad-auth-type",
+			split("config add-user --username foobar --auth-type badbad")},
+	}
+
+	if len(encryptPassword) > 2 { // are we on a platform that supports encryption
+		tests = append(tests, test{"neg-config-add-user-bad-use-encrypted",
+			split(fmt.Sprintf("config add-user --username foobar --auth-type other%v", encryptPassword))})
+	}
+
+	run(t, tests)
+}
+
+func TestGetTags(t *testing.T) {
+	setup(t.Name())
+	tests := []test{
+		{"get-tags",
+			split("install mssql get-tags")},
+	}
+
+	run(t, tests)
+}
+
+func TestMssqlInstall(t *testing.T) {
+	setup(t.Name())
+	tests := []test{
+		{"install",
+			split(fmt.Sprintf("install mssql%v --user-database my-database --accept-eula%v", useCached, encryptPassword))},
+		{"config-current-context",
+			split("config current-context")},
+		{"config-connection-strings",
+			split("config connection-strings")},
+		{"query",
+			split("query GO")},
+		{"query",
+			split("query")},
+		{"neg-query-two-queries",
+			split("query bad --query bad")},
+
+		/* How to get code coverage for user input
+		{"neg-uninstall-no-yes",
+			split("uninstall")},*/
+		{"uninstall",
+			split("uninstall --yes --force")},
+	}
+
+	run(t, tests)
+}
+
+func runTests(t *testing.T, tt struct {
+	name string
+	args struct{ args []string }
+}) {
+	cmd := cmdparser.New[*Root](root.SubCommands()...)
+	cmd.ArgsForUnitTesting(tt.args.args)
+
+	t.Logf("Running: %v", tt.args.args)
+
+	if tt.name == "neg-config-add-user-no-password" {
+		os.Setenv("SQLCMD_PASSWORD", "")
+	} else {
+		os.Setenv("SQLCMD_PASSWORD", "badpass")
+	}
+
+	// If test name starts with 'neg-' expect a Panic
+	if strings.HasPrefix(tt.name, "neg-") {
+		defer func() {
+			if r := recover(); r == nil {
+				t.Errorf("The code did not panic")
+			}
+		}()
+		cmd.Execute()
+	}
+	cmd.Execute()
```

Comment:
is silence considered success? The lack of asserts makes it hard to tell what's being tested.

Reply:
Agree, I was focusing on getting to 100% code coverage, but now I need to go back and actually validate the test results.  To keep the size of the PR managable I'll follow up now with a Test focused PR.